File: Demo-LLP-MIP/Demo-LLP-MIP-PHP/security/LLianPayMipSignature.php
<?php

require_once 'D:\file\lianlian\code\Demo-LLP-MIP-PHP\vendor\autoload.php';
use phpseclib\Crypt\RSA;
require_once 'D:\file\lianlian\code\Demo-LLP-MIP-PHP\vendor\phpseclib\phpseclib\phpseclib\Crypt\RSA.php';
require_once 'D:\file\lianlian\code\Demo-LLP-MIP-PHP\config\LLianPayConstant.php';
require_once 'D:\file\lianlian\code\Demo-LLP-MIP-PHP\security\Base64.php';

class LLianPayMipSignature {
    public function sign($sign_str, $prikeyvalue=null) {
        if ($prikeyvalue === null) {
            include 'D:\file\lianlian\code\Demo-LLP-MIP-PHP\config\LLianPayConstant.php';
            $prikeyvalue = $MerchantPrivateKey;
        }
        echo $sign_str . "\n";
        $hash_value = hash('md5', $sign_str);
        echo "签名处理中，签名源内容：" . $sign_str . "，对应MD5值：" . $hash_value;
        return $this->RSA_sign($prikeyvalue, $hash_value);
    }

    function RSA_sign($prikeyvalue, $sign_str) {
        try {
            $prikeyDer = base64_decode($prikeyvalue);
            $pem = "-----BEGIN PRIVATE KEY-----\n" . chunk_split(base64_encode($prikeyDer), 64, "\n") . "-----END PRIVATE KEY-----";

            $pkeyid = openssl_pkey_get_private($pem);
            if (!$pkeyid) {
                throw new Exception("Failed to extract private key");
            }

            // 用私钥对信息生成数字签名
            openssl_sign($sign_str, $signature, $pkeyid, OPENSSL_ALGO_MD5);
            openssl_free_key($pkeyid);

            return base64_encode($signature);
        } catch (Exception $e) {
            error_log("签名失败," . $e->getMessage());
        }
        return null;
    }

    public function checkSign($sign_str, $signed_str, $pubkeyvalue=null) {
        if ($pubkeyvalue === null) {
            include 'D:\file\lianlian\code\Demo-LLP-MIP-PHP\config\LLianPayConstant.php';
            $pubkeyvalue = $LLianPayPublicKey;
        }

        $hash_value = hash('md5', $sign_str);
        echo "\n签名验证中，源串：" . $sign_str . "，对应MD5值：" . $hash_value;
        return $this->RSA_checksign($pubkeyvalue, $hash_value, $signed_str);
    }

    function RSA_checksign($pubkeyvalue, $hash_value, $signed_str) {
        try {
            // 将Base64编码的公钥值转换为字节形式
            $pubkeyvalue_bytes = base64_decode($pubkeyvalue);
            $public_key = openssl_pkey_get_public("-----BEGIN PUBLIC KEY-----\n" . chunk_split(base64_encode($pubkeyvalue_bytes), 64, "\n") . "-----END PUBLIC KEY-----");
            if ($public_key === false) {
                throw new Exception("Failed to extract public key");
            }

            // 创建一个用于验证签名的 OpenSSL key 对象
            $rsaObj = openssl_verify($hash_value, base64_decode($signed_str), $public_key, OPENSSL_ALGO_MD5);
            if ($rsaObj == 1) {
                return true;
            } elseif ($rsaObj == 0) {
                return false;
            } else {
                throw new Exception("An error occurred while verifying the signature");
            }
        } catch (Exception $e) {
            error_log("签名验证异常," . $e->getMessage());
        }
        return false;
    }

    public function localEncrypt($sourceStr) {
        try {
            $rsa = new RSA();
            $publicKey = LLianPayPublicKey;
            $rsa->loadKey(Base64::decode($publicKey));
            $rsa->setEncryptionMode(RSA::ENCRYPTION_PKCS1);
            $encryptStr = $rsa->encrypt($sourceStr);
            $encryptStrBase64 = base64_encode($encryptStr);
            echo "本地RSA加密，源串：" . $sourceStr . "，加密后值：" . $encryptStr;
            return $encryptStrBase64;
        } catch (Exception $e) {
            echo "本地RSA加密异常, " . $e;
        }
    }
}
?>

File: Demo-LLP-MIP/Demo-LLP-MIP-Python/security/LLianPayMipSignature.py
import logging
log = logging.getLogger('LLianPayMipSignature')
log.setLevel(logging.INFO)
import hashlib
from Crypto.Signature import PKCS1_v1_5
from Crypto.Hash import MD5
from Crypto.Cipher import PKCS1_v1_5 as Cipher_PKCS1_v1_5
from Crypto.PublicKey import RSA
from config.LLianPayConstant import MerchantPrivateKey
from config.LLianPayConstant import LLianPayPublicKey
from security.Base64 import *

# 定义一个LLianPayMipSignature类，实现连连账户+签名功能
class LLianPayMipSignature:
    def __init__(self):
        pass

    @staticmethod
    def getInstance():
        return LLianPayMipSignature()

    # 签名处理，默认使用LLianPayConstant中配置的私钥MerchantPrivateKey
    def sign(self, sign_str, prikeyvalue=None):
        if prikeyvalue is None:
            prikeyvalue = MerchantPrivateKey
        logging.info(sign_str)
        hash_value = hashlib.md5(sign_str.encode('utf-8')).hexdigest()
        log.info("签名处理中，签名源内容：%s，对应MD5值：%s" % (sign_str, hash_value))
        return self.RSA_sign(prikeyvalue, hash_value)

    # RSA签名
    def RSA_sign(self, prikeyvalue, hash_value):
        try:
            # prikeyvalue 是私钥的值
            # 已确认相同
            prikeyvalue_bytes = Base64.get_bytes_from_base64(prikeyvalue)

            # 使用私钥值创建RSA私钥对象
            private_key = RSA.import_key(prikeyvalue_bytes)
            # 用私钥对信息生成数字签名
            signer = PKCS1_v1_5.new(private_key)
            hash_value = MD5.new(hash_value.encode('utf-8'))
            signature = signer.sign(hash_value)

            # 将签名结果进行Base64编码
            signed_base64 = base64.b64encode(signature).decode('utf-8')

            # 返回Base64编码后的签名结果
            return signed_base64
        except Exception as e:
            log.error("签名失败,%s" % e)

    # 签名验证，默认使用LLianPayConstant中配置的公钥LLianPayPublicKey
    def checkSign(self, sign_str, signed_str, pubkeyvalue=None):
        if pubkeyvalue is None:
            pubkeyvalue = LLianPayPublicKey
        hash_value = hashlib.md5(sign_str.encode('utf-8')).hexdigest()
        log.info("签名验证中，源串：%s，对应MD5值：%s" % (sign_str, hash_value))
        return self.RSA_checksign(pubkeyvalue, hash_value, signed_str)

    # RSA签名验证
    def RSA_checksign(self, pubkeyvalue, hash_value, signed_str):
        try:
            # 将Base64编码的公钥值转换为字节形式
            pubkeyvalue_bytes = base64.b64decode(pubkeyvalue)
            public_key = RSA.importKey(pubkeyvalue_bytes)
            cipher = PKCS1_v1_5.new(public_key)
            h = MD5.new(hash_value.encode('utf-8'))
            signature = base64.b64decode(signed_str)
            return cipher.verify(h, signature)
        except Exception as e:
            log.error("签名验证异常,%s" % e)

    # 测试环境使用连连公钥加密密码
    def localEncrypt(self, sourceStr):
        try:
            public_key = RSA.import_key(LLianPayPublicKey)
            cipher_rsa = Cipher_PKCS1_v1_5.new(public_key)
            encrypt_bytes = cipher_rsa.encrypt(sourceStr.encode('utf-8'))
            encryptStr = base64.b64encode(encrypt_bytes).decode('utf-8')
            log.info("本地RSA加密，源串：%s，加密后值：%s" % (sourceStr, encryptStr))
            return encryptStr
        except Exception as e:
            log.error("本地RSA加密异常,%s" % e)